wallet-api/src/Wallet/Emulator/Http.hs

```diff
@@ -0,0 +1,164 @@
+  if HM.member walletId ws
+    then pure (Wallet walletId)
+    else throwError err404
+
+createWallet :: (MonadReader State m, MonadIO m) => Wallet -> m NoContent
+createWallet wallet = do
+  var <- asks getWallets
+  let walletState = emptyWalletState wallet
+  liftIO . atomically $
+    modifyTVar var (HM.insert (getWallet wallet) walletState)
+  pure NoContent
+
+createPaymentWithChange ::
+     (MonadReader State m, MonadIO m, MonadError ServantErr m)
+  => Int
+  -> Value
+  -> m (Set.Set TxIn', TxOut')
+createPaymentWithChange = runWalletApiAction WAPI.createPaymentWithChange
+
+payToPublicKey ::
+     (MonadReader State m, MonadIO m, MonadError ServantErr m)
+  => Int
+  -> Value
+  -> m TxOut'
+payToPublicKey = runWalletApiAction WAPI.payToPublicKey
+
+submitTxn ::
+     (MonadReader State m, MonadIO m, MonadError ServantErr m)
+  => Int
+  -> Tx
+  -> m ()
+submitTxn = runWalletApiAction WAPI.submitTxn
+
+runWalletApiAction ::
+     (MonadReader State m, MonadIO m, MonadError ServantErr m)
+  => (a -> EmulatedWalletApi b)
+  -> Int
+  -> a
+  -> m b
+runWalletApiAction action walletId a = do
+  var <- asks getWallets
+  finalResult <-
+    liftIO . atomically $ do
+      states <- readTVar var
+      let mState = HM.lookup walletId states
+      case mState of
+        Nothing -> pure . Left . HttpError $ err404
+        Just oldState -> do
+          let (res, _) =
+                runWriter .
+                flip runStateT oldState . runExceptT . runEmulatedWalletApi $
+                action a
+          case res of
+            (Left e, _) -> pure . Left . WalletAPIError $ e
+            (Right v, newState) -> do
+              modifyTVar var (HM.insert walletId newState)
+              pure $ Right v
+  case finalResult of
+    Left (HttpError e) -> throwError e
+    Left (WalletAPIError e) ->
+      throwError $ err500 {errBody = BSL.pack . show $ e}
+    Right res -> pure res
+
+-- TODO: This will be part of the transactions API
+getTransactions ::
+     (MonadReader State m, MonadIO m, MonadError ServantErr m) => m [Tx]
+getTransactions = undefined
+
+-- | Concrete monad stack for server server
+newtype M a = M
+  { unM :: ReaderT State (ExceptT ServantErr IO) a
+  } deriving ( Functor
+             , Applicative
+             , Monad
+             , MonadReader State
+             , MonadIO
+             , MonadError ServantErr
+             )
+
+runM :: State -> M ~> Handler
+runM state r = do
+  res <- liftIO . runExceptT . flip runReaderT state . unM $ r
+  case res of
+    Right response -> pure response
+    Left e         -> throwError e
+
+walletHandlers :: State -> Server WalletAPI
+walletHandlers state =
+  hoistServer walletApi (runM state) $
+  wallets :<|> fetchWallet :<|> createWallet :<|> createPaymentWithChange :<|>
+  payToPublicKey :<|>
+  submitTxn :<|>
+  getTransactions
+
+walletApi :: Proxy WalletAPI
+walletApi = Proxy
+
+app :: State -> Application
+app state = serve walletApi $ walletHandlers state
```

Comment:
OOI is there any particular reason not to use `TypeApplications` and do `Proxy @WalletApi`?

Reply:
meh, on the fence as to which looks better, I will change it if you feel its worth it

Reply:
I would tend to do the inline one, but it doesn't matter much. Just wondering if there was a reason it wouldn't work.

Reply:
@michaelpj Nice. Hadn't occurred to me that you could use TypeApplications for that. :-)

Reply:
ah, I was just about to change it when I saw that it is used in 2 places, I think I prefer to have this than to lots of `(Proxy @WalletAPI)`